src/jattach/fdtransfer_server.cpp

```diff
@@ -0,0 +1,269 @@
+                error = ESRCH;
+            }
+
+            struct perf_fd_response resp;
+            resp.header.response_id = request->header.request_id;
+            resp.header.error = error;
+            resp.tid = request->tid;
+            sendFd(perf_fd, &resp.header, sizeof(resp));
+            close(perf_fd);
+            break;
+        }
+
+        case KALLSYMS_FD: {
+            // can't directly pass the fd of /proc/kallsyms, because before Linux 4.15 the permission check
+            // was conducted on each read.
+            // it's simpler to copy the file to a temporary location and pass the fd of it (compared to passing the
+            // entire contents over the peer socket)
+            char tmp_path[256];
+            snprintf(tmp_path, sizeof(tmp_path), TMP_KALLSYMS_PATH, getpid());
+            {
+                // TODO from Linux 3.15, we can use memfds here. Also, from 3.11 we can use O_TMPFILE.
+                std::ifstream src("/proc/kallsyms", std::ios::binary);
+                close(creat(tmp_path, S_IRUSR | S_IWUSR));  // create it with 0600 so others can't read.
+                std::ofstream dst(tmp_path, std::ios::binary);
+                dst << src.rdbuf();
+            }
+
+            struct fd_response resp;
+            resp.response_id = req->request_id;
+            int kallsyms_fd = open(tmp_path, O_RDONLY);
+            if (kallsyms_fd == -1) {
+                resp.error = errno;
+            } else {
+                unlink(tmp_path);
+                resp.error = 0;
+            }
+
+            sendFd(kallsyms_fd, &resp, sizeof(resp));
+            close(kallsyms_fd);
+            break;
+        }
+
+        default:
+            fprintf(stderr, "unknown request type %u\n", req->type);
+            break;
+        }
+    }
+
+    return false;
+}
+
+bool FdTransferServer::sendFd(int fd, struct fd_response *resp, size_t resp_size) {
+    struct msghdr msg = {0};
+
+    struct iovec iov[1];
+    iov[0].iov_base = resp;
+    iov[0].iov_len = resp_size;
+    msg.msg_iov = iov;
+    msg.msg_iovlen = ARRAY_SIZE(iov);
+
+    union {
+       char buf[CMSG_SPACE(sizeof(fd))];
+       struct cmsghdr align;
+    } u;
+
+    if (fd != -1) {
+        msg.msg_control = u.buf;
+        msg.msg_controllen = sizeof(u.buf);
+
+        struct cmsghdr *cmsg = CMSG_FIRSTHDR(&msg);
+        cmsg->cmsg_level = SOL_SOCKET;
+        cmsg->cmsg_type = SCM_RIGHTS;
+        cmsg->cmsg_len = CMSG_LEN(sizeof(fd));
+        memcpy(CMSG_DATA(cmsg), &fd, sizeof(fd));
+    }
+
+    ssize_t ret = sendmsg(_peer, &msg, 0);
+    if (ret < 0) {
+        perror("sendmsg()");
+        return false;
+    }
+
+    return true;
+}
+
+static bool enter_pid_net_mnt(pid_t pid) {
+    // the last one works - because we still have the old /proc accessible, so we see PIDs as they are
+    // the NS before we moved.
+    return enter_ns(pid, "net") == 1 && enter_ns(pid, "pid") == 1 && enter_ns(pid, "mnt");
```

Comment:
Wait a minute... why network namespace?
IIUC, to access a named UNIX socket you need to be in the same `mnt` namespace. `net` makes difference only for abstract UNIX sockets.

Reply:
Replied in https://github.com/jvm-profiling-tools/async-profiler/pull/411#issuecomment-894647869